corehq/util/couch_doc_processor.py

```diff
@@ -0,0 +1,372 @@
+import hashlib
+from abc import abstractmethod, ABCMeta, abstractproperty
+from datetime import datetime
+
+import six
+from couchdbkit import ResourceNotFound
+
+from corehq.util.couch_helpers import PaginateViewLogHandler, paginate_view
+
+
+class ResumableDocsByTypeIterator(object):
+    """Perform one-time resumable iteration over documents by type
+
+    Iteration can be efficiently stopped and resumed. The iteration may
+    omit documents that are added after the iteration begins or resumes
+    and may include deleted documents.
+
+    Example usage
+    > iterator = ResumableDocsByTypeIterator(db, doc_types, 'my-key')
+    > for doc in iterator:
+        success = process_doc(doc)
+        if success:
+            iterator.mark_doc_processed(doc)
```

Comment:
I'm not understanding the need for this. What does it add that was not previously possible? This method does nothing in the typical case, and makes the retry logic more complicated. If I was reading this for the first time I would think it is necessary to call `iterator.mark_doc_processed(doc)` even if I was not using `retry` (not true). The intuitive way to mark a document as processed is to proceed to the next item in the iteration without calling `iterator.retry(doc)`.

Maybe we should back up: do you have a use case in mind that is not served by the simple retry mechanism originally in `ResumableDocsByTypeIterator`?